### ISSUE TITLE
Add Data Layer

### ISSUE THREAD

**Issue by dolfelt** [dolfelt/proton-example#1]
Figure out what kind of data layer to add, and which library to use.
- [Spot ORM](https://github.com/vlucas/spot2)
- [Propel](https://github.com/propelorm/Propel2)
- Doctrine, etc.

I'm liking the DBAL concept, which is what Spot ORM is based on, where you have Entities and Mappers to handle the full connection to the data store.

cc: @shadowhand @ameech

**Comment by shadowhand** [comment]
I'll throw another library into the mix: [JSONModel](https://github.com/unframed/JSONModel.php), but...

From my perspective, deciding on database details is a last step for planning the app, not the first. I'd much rather we focus on how data flows through the system by talking about design patterns before specifics about tools. Here's some fodder for discussion:

https://en.wikipedia.org/wiki/Delegation_pattern
https://en.wikipedia.org/wiki/Command%E2%80%93query_separation
https://en.wikipedia.org/wiki/Command_pattern
https://en.wikipedia.org/wiki/State_pattern
https://en.wikipedia.org/wiki/Strategy_pattern

These are not mutually exclusive and we can choose the right patterns for the right parts of the application.

In my mind, the purpose of an example app should be to describe the types of classes that exist, where those classes live in the filesystem, and how to implement a new input/output stream within the application. A good starting point for such a discussion is [Clean Architecture](http://blog.8thlight.com/uncle-bob/2012/08/13/the-clean-architecture.html) and exploration of how a REST API can be effectively modeled using hexagonal architecture.

**Comment by dolfelt** [comment]
I agree. Just throwing this out there as a discussion starter. This example is far from complete. Mostly I'm trying to get a logical framework for a large application.

I like the delegation pattern a lot and think we could accomplish most of the "clean architecture" using that.
